src/clusterfuzz/_internal/system/archive.py

```diff
@@ -34,111 +36,316 @@
 
-  if not file_match_callback:
+@dataclasses.dataclass
+class ArchiveMemberInfo:
+  """Represents an archive member.
+  """
+  filename: str
+  is_dir: bool
+  file_size_bytes: int
+  mode: int
+
 
-    def file_match_callback_fallback(_):
-      return True
+class ArchiveReader(abc.ABC):
+  """Abstract class for representing an archive reader. Abstract methods must
+  be overriden.
+  """
 
-    file_match_callback = file_match_callback_fallback
+  @abc.abstractmethod
+  def list_files(self) -> typing.List[ArchiveMemberInfo]:
+    """Lists all members contained in the archives.
 
-  def maybe_extract(extract_func, info):
-    """Returns an extracted file or None if it is not supposed to be extracted.
+    Returns:
+        typing.List[ArchiveMemberInfo]: All the archive members
     """
-    if should_extract:
-      return extract_func(info)
-    return None
+    raise NotImplementedError
 
-  if archive_type == ArchiveType.ZIP:
-    try:
-      with zipfile.ZipFile(archive_obj or archive_path) as zip_file:
-        for info in zip_file.infolist():
-          if not file_match_callback(info.filename):
-            continue
+  @abc.abstractmethod
+  def extract(self, member, path, trusted=False) -> str:
+    """Extracts `member` out of the archive to the provided path.
+
+    Args:
+        member (str): the member name
+        path (str): the path where the member should be extracted.
+        Defaults to None.
+        trusted (bool, optional): whether the archive is trusted. Defaults to
+        False.
+
+    Returns:
+        str: the path to the extracted member
+    """
+    raise NotImplementedError
+
+  @abc.abstractmethod
+  def open(self, member):
+    """Opens `member`.
+
+    Args:
+        member (str): the member name
+
+    Returns:
+        [IO]: a file-like object that can be `read`.
+    """
+    raise NotImplementedError
+
+  @abc.abstractmethod
+  def close(self) -> None:
+    """Closes the archive.
+    """
+    raise NotImplementedError
+
+  @abc.abstractmethod
+  def extractall(self, path, members=None, trusted=False) -> None:
+    """Extract the whole archive content or the members listed in `members`.
+
+    Args:
+        path (str): the path where the members should be extracted.
+        Defaults to None.
+        members ([str], optional): the member names. Defaults to None.
+        trusted (bool, optional): whether the archive is trusted or not.
+        Defaults to False.
+    """
+    raise NotImplementedError
 
-          yield ArchiveFile(info.filename, info.file_size,
-                            maybe_extract(zip_file.open, info))
+  def try_open(self, member):
+    """Tries to open the archive. Does not throw.
 
-    except (zipfile.BadZipfile, zipfile.LargeZipFile):
-      logs.log_error('Bad zip file %s.' % archive_path)
+    Args:
+        member (str): the member name
 
-  elif archive_type == ArchiveType.TAR:
+    Returns:
+        [IO]: a file-like object that can be `read`.
```

Comment:
Or `None`. The function can be annotated as `-> Optional[IO]` (with `from typing import Optional`).